src/equations/serre_green_naghdi_1d.jl

```diff
@@ -0,0 +1,960 @@
+        @. tmp = v^2
+        mul!(v2_x, D, tmp)
+
+        @. tmp = h^2 * v * v_x
+        mul!(h2_v_vx_x, D, tmp)
+        @. tmp = h * v_x
+        mul!(h_vx_x, D, tmp)
+        inv6 = 1 / 6
+        @. tmp = (0.5 * h^2 * (h * v_x + h_x * v) * v_x
+                  -
+                  inv6 * h * h2_v_vx_x
+                  -
+                  inv6 * h^2 * v * h_vx_x)
+        mul!(p_x, D, tmp)
+
+        # Plain: h_t + (h v)_x = 0
+        #
+        # Split form for energy conservation:
+        # h_t + h_x v + h v_x = 0
+        @. dh = -(h_x * v + h * v_x)
+
+        # Plain: h v_t + ... = 0
+        #
+        # Split form for energy conservation:
+        @. tmp = -(g * h2_x - g * h * h_x
+                   +
+                   0.5 * h * v2_x
+                   -
+                   0.5 * v^2 * h_x
+                   +
+                   0.5 * hv_x * v
+                   -
+                   0.5 * h * v * v_x
+                   +
+                   p_x)
+    end
+
+    @trixi_timeit timer() "assembling elliptic operator" begin
+        # The code below is equivalent to
+        #   dv .= (Diagonal(h) - Dmat * Diagonal(1/3 .* h.^3) * Dmat) \ tmp
+        # but faster since the symbolic factorization is reused.
+        # Floating point errors accumulate a bit and the system matrix
+        # is not necessarily perfectly symmetric but only up to round-off
+        # errors. We wrap it here to avoid issues with the factorization.
+        @. M_h = h
+        scale_by_mass_matrix!(M_h, D)
+        inv3 = 1 / 3
+        @. M_h3_3 = inv3 * h^3
+        scale_by_mass_matrix!(M_h3_3, D)
+        system_matrix = Symmetric(Diagonal(M_h)
+                                  +
+                                  Dmat' * Diagonal(M_h3_3) * Dmat)
+    end
+
+    @trixi_timeit timer() "solving elliptic system" begin
+        if issparse(system_matrix)
+            (; factorization) = cache
+            cholesky!(factorization, system_matrix; check = false)
+            if issuccess(factorization)
+                scale_by_mass_matrix!(tmp, D)
+                dv .= factorization \ tmp
+            else
+                # The factorization may fail if the time step is too large
+                # and h becomes negative.
+                fill!(dv, NaN)
+            end
+        else
+            factorization = cholesky!(system_matrix)
+            scale_by_mass_matrix!(tmp, D)
+            ldiv!(dv, factorization, tmp)
+        end
+    end
+
+    return nothing
+end
+
+function rhs_sgn_upwind!(dq, q, equations, source_terms, cache, ::BathymetryFlat)
+    # Unpack physical parameters and SBP operator `D` as well as the
+    # SBP upwind operator in sparse matrix form `Dmat_minus`
+    g = gravity_constant(equations)
+    (; Dmat_minus) = cache
+    D_upwind = cache.D
+    D = D_upwind.central
+
+    # `q` and `dq` are `ArrayPartition`s. They collect the individual
+    # arrays for the water height `h` and the velocity `v`.
+    h, v = q.x
```

Comment:
```suggestion
    eta, v = q.x
```
This would be `\eta` usually. Of course the remaining part of this function needs to be adjusted, too.